omp/components/prefix_sum.cpp

```diff
@@ -33,22 +33,67 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include "core/components/prefix_sum.hpp"
 
 
+#include <algorithm>
+#include <vector>
+
+
+#include <omp.h>
+
+
 namespace gko {
 namespace kernels {
 namespace omp {
 namespace components {
 
 
+/*
+ * The last entry of the input array is never used, but is replaced.
+ */
 template <typename IndexType>
-void prefix_sum(std::shared_ptr<const OmpExecutor> exec, IndexType *counts,
-                size_type num_entries)
+void prefix_sum(std::shared_ptr<const OmpExecutor> exec,
+                IndexType *const counts, const size_type num_entries)
 {
-    IndexType partial_sum{};
-    for (size_type i = 0; i < num_entries; ++i) {
-        auto nnz = counts[i];
-        counts[i] = partial_sum;
-        partial_sum += nnz;
+    const auto nentries = static_cast<IndexType>(num_entries);
+    // the operation only makes sense for arrays of size at least 2
+    GKO_ENSURE_COMPATIBLE_BOUNDS(2, nentries);
+
+    const int nthreads = omp_get_max_threads();
+    std::vector<IndexType> proc_sums(nthreads, 0);
```

Comment:
We could use a custom allocator here, allowing us to track the allocations through loggers
```suggestion
    vector<IndexType> proc_sums(nthreads, 0, {exec});
```